metricflow/time/time_granularity_solver.py

```diff
@@ -253,6 +268,80 @@ def adjust_time_range_to_granularity(
 
         return TimeRangeConstraint(start_time=constraint_start, end_time=constraint_end)
 
+    def _get_primary_time_dimension(self) -> TimeDimensionSpec:  # noqa: D
+        for data_source in self._semantic_model.user_configured_model.data_sources:
+            for dimension in data_source.dimensions:
+                if dimension.type_params:
+                    if dimension.type_params.time_granularity:
+                        if dimension.type_params.is_primary:
+                            return TimeDimensionSpec.parse(dimension.name.element_name)
+        # Shouldn't happen as validation should have caught this case.
+        raise RuntimeError("No primary time dimension found in the model.")
+
+    def _get_greatest_common_time_granularity(self, metrics: Sequence[MetricSpec]) -> TimeGranularity:
+        """Return the greatest common granularity for metrics"""
+        primary_time_dimension_reference = self._get_primary_time_dimension().reference
+        _, max_time_granularity = self.local_dimension_granularity_range(
+            metric_specs=metrics, local_time_dimension_reference=primary_time_dimension_reference
+        )
+        return max_time_granularity
+
+    def _have_consistent_time_granularity(
+        self, metric_specs: Sequence[MetricSpec], local_time_dimension_reference: TimeDimensionReference
+    ) -> bool:
+        """Returns whether all of the measures in all of the metrics have the same time granularity"""
+        min_time_granularity, max_time_granularity = self.local_dimension_granularity_range(
+            metric_specs, local_time_dimension_reference
+        )
+        return min_time_granularity == max_time_granularity
+
+    def solve_time_granularity_transformation(
+        self,
+        metric_specs: Sequence[MetricSpec],
+        time_dimension_specs: Sequence[TimeDimensionSpec],
+    ) -> TimeGranularityValidationReference:
+        """Figure out how to transform the time granularity in a query."""
+
+        primary_time_dimension_spec: TimeDimensionSpec = self._get_primary_time_dimension()
+        min_time_granularity_for_querying = self._get_greatest_common_time_granularity(metrics=metric_specs)
+
+        requested_time_granularity = None
+        for time_dimension_spec in time_dimension_specs:
+            if primary_time_dimension_spec == time_dimension_spec:
+                if time_dimension_spec.time_granularity < min_time_granularity_for_querying:
```

Comment:
I am confused about these nested conditionals. Why are we only checking the primary time dimension for the model? Also, I always get confused by this and it's something we should address. I think the issue here is that we've pre-materialized and now the lower granularity is effectively a request for a granularity we can no longer serve, is that correct? If so, we should probably find a way to document this more clearly, because in this location in the codebase it's not at all obvious why a lower granularity on the primary time dimension is an issue.

Also worth noting, if this is targeted at model validation, it should probably be separated out from the solution steps and this method should be targeted more directly at model validation outcomes. If this is targeted at finding a granularity transformation, I'd expect this logic to fail if any requested time dimension was incorrect.